### PR TITLE
[SPARK-46731][SS] Manage state store provider instance by state data source - reader

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to change state data source - reader part to manage state store provider instance by itself.

### Why are the changes needed?

Currently, state data source initializes state store instance via StateStore.get() which also initializes state store provider instance and registers the provider instance to the coordinator. This involves unnecessary overheads e.g. maintenance task could be triggered for this provider.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @viirya @anishshri-db Please take a look. Thanks in advance!

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @viirya @anishshri-db Friendly reminder.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala [hunk 1]
```diff
@@ -443,6 +444,7 @@ class SymmetricHashJoinStateManager(
 
   /** Helper trait for invoking common functionalities of a state store. */
   private abstract class StateStoreHandler(stateStoreType: StateStoreType) extends Logging {
+    private var stateStoreProvider: StateStoreProvider = _
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:447]: Hmm, is this change related to state data source reader?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:447]: Oh, nvm, I think I know where this is used now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala [hunk 2]
```diff
@@ -79,7 +79,8 @@ class SymmetricHashJoinStateManager(
     hadoopConf: Configuration,
     partitionId: Int,
     stateFormatVersion: Int,
-    skippedNullValueCount: Option[SQLMetric] = None) extends Logging {
+    skippedNullValueCount: Option[SQLMetric] = None,
+    useStateStoreCoordinator: Boolean = true) extends Logging {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:83]: Could you update this param into the class doc?

**Review by viirya — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:447]: Yeah that is actually tricky but there does not seem to be any alternative.

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Thanks for reviewing! Merging to master.
